File: ebuild/private/common/portage/binarypackage/binarypackage.go
// Copyright 2022 The ChromiumOS Authors.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package binarypackage

import (
	"encoding/binary"
	"errors"
	"fmt"
	"io"
	"os"
	"os/exec"
	"syscall"
)

// See https://www.mankier.com/5/xpak for the format specification.
type File struct {
	xpakStart int64
	size      int64
	f         *os.File
}

type XPAK map[string][]byte

func BinaryPackage(path string) (bp *File, err error) {
	bp = &File{}
	bp.f, err = os.Open(path)
	if err != nil {
		return nil, err
	}

	success := false
	defer func() {
		if !success {
			bp.Close()
		}
	}()

	fi, err := bp.f.Stat()
	if err != nil {
		return nil, err
	}
	bp.size = fi.Size()

	if bp.size < 24 {
		return nil, errors.New("corrupted .tbz2 file: size is too small")
	}
	if err := bp.expectMagic(bp.size-4, "STOP"); err != nil {
		return nil, fmt.Errorf("corrupted .tbz2 file: %w", err)
	}
	xpakOffset, err := bp.readUint32(bp.size - 8)
	if err != nil {
		return nil, fmt.Errorf("corrupted .tbz2 file: %w", err)
	}
	bp.xpakStart = bp.size - 8 - int64(xpakOffset)
	if bp.xpakStart < 0 {
		return nil, errors.New("corrupted .tbz2 file: invalid xpak_offset")
	}
	if err := bp.expectMagic(bp.size-16, "XPAKSTOP"); err != nil {
		return nil, fmt.Errorf("corrupted .tbz2 file: %w", err)
	}
	if err := bp.expectMagic(bp.xpakStart, "XPAKPACK"); err != nil {
		return nil, fmt.Errorf("corrupted .tbz2 file: %w", err)
	}

	success = true
	return bp, nil
}

func (bp *File) Close() error {
	return bp.f.Close()
}

func (bp *File) TarballReader() (io.ReadCloser, error) {
	newFd, err := syscall.Dup(int(bp.f.Fd()))
	if err != nil {
		return nil, err
	}
	f := os.NewFile(uintptr(newFd), bp.f.Name())
	if _, err := f.Seek(0, io.SeekStart); err != nil {
		return nil, err
	}
	return readCloser{
		Reader: io.LimitReader(f, bp.xpakStart),
		Closer: f,
	}, nil
}

func (bp *File) Merge(path string) error {
	tarball, err := bp.TarballReader()
	if err != nil {
		return err
	}
	defer tarball.Close()

	// Note that at the moment, ownership is not retained.
	// --same-owner should fix that, but:
	// 1) it can only be run with sudo.
	// 2) If we generate bazel output files with strange ownership, bazel won't
	// have permissions to clean it up.
	// When we write to an image, we'll need to do some work to preserve metadata.
	cmd := exec.Command("tar", "--zstd", "--keep-old-files", "--same-permissions", "-xf", "-")
	cmd.Dir = path
	cmd.Stdin = tarball
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	if err := cmd.Run(); err != nil {
		return fmt.Errorf("failed to extract from %s - maybe multiple packages attempt to define the same file: %w", bp.f.Name(), err)
	}
	return nil
}

func (bp *File) Xpak() (XPAK, error) {
	indexLen, err := bp.readUint32(bp.xpakStart + 8)
	if err != nil {
		return nil, err
	}
	dataLen, err := bp.readUint32(bp.xpakStart + 12)
	if err != nil {
		return nil, err
	}
	indexStart := bp.xpakStart + 16
	dataStart := indexStart + int64(indexLen)
	if dataStart+int64(dataLen) != bp.size-16 {
		return nil, fmt.Errorf("corrupted .tbz2 file: data length inconsistency")
	}

	xpak := make(map[string][]byte)
	for indexPos := indexStart; indexPos < dataStart; {
		nameLen, err := bp.readUint32(indexPos)
		if err != nil {
			return nil, err
		}
		indexPos += 4
		nameBuf := make([]byte, int(nameLen))
		if _, err := io.ReadFull(bp.f, nameBuf); err != nil {
			return nil, err
		}
		indexPos += int64(nameLen)
		name := string(nameBuf)
		dataOffset, err := bp.readUint32(indexPos)
		if err != nil {
			return nil, err
		}
		indexPos += 4
		dataLen, err := bp.readUint32(indexPos)
		if err != nil {
			return nil, err
		}
		indexPos += 4

		if _, err := bp.f.Seek(dataStart+int64(dataOffset), io.SeekStart); err != nil {
			return nil, err
		}
		data := make([]byte, int(dataLen))
		if _, err := io.ReadFull(bp.f, data); err != nil {
			return nil, err
		}

		xpak[name] = data
	}

	return xpak, nil
}

func (bp *File) readUint32(offset int64) (uint32, error) {
	if _, err := bp.f.Seek(offset, io.SeekStart); err != nil {
		return 0, err
	}
	buf := make([]byte, 4)
	if _, err := io.ReadFull(bp.f, buf); err != nil {
		return 0, err
	}
	return binary.BigEndian.Uint32(buf), nil
}

func (bp *File) expectMagic(offset int64, want string) error {
	if _, err := bp.f.Seek(offset, io.SeekStart); err != nil {
		return err
	}
	buf := make([]byte, len(want))
	if _, err := io.ReadFull(bp.f, buf); err != nil {
		return err
	}
	if got := string(buf); got != want {
		return fmt.Errorf("bad magic: got %q, want %q", got, want)
	}
	return nil
}

func ReadXpak(path string) (XPAK, error) {
	bp, err := BinaryPackage(path)
	if err != nil {
		return nil, err
	}
	defer bp.Close()
	return bp.Xpak()
}

type readCloser struct {
	io.Reader
	io.Closer
}

File: ebuild/private/common/portage/binarypackage/src/binarypackage.rs
// Copyright 2022 The ChromiumOS Authors.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

use anyhow::{bail, Context, Result};
use bytes::ByteOrder;
use std::{
    collections::HashMap,
    fs::File,
    io::SeekFrom::Start,
    io::{Read, Seek},
    os::unix::fs::MetadataExt,
    path::Path,
};

/// Works with Portage binary package files (.tbz2).
///
/// See https://www.mankier.com/5/xpak for the format specification.
pub struct BinaryPackage {
    file: File,
    xpak_start: u64,
    xpak: HashMap<String, Vec<u8>>,
    category_pf: String,
}

impl BinaryPackage {
    /// Opens a Portage binary package file.
    pub fn open(path: &Path) -> Result<Self> {
        let mut file = File::open(path)?;
        let metadata = std::fs::metadata(path)?;
        let size = metadata.size();

        if size < 24 {
            bail!("corrupted .tbz2 file: size is too small")
        }

        const CORRUPTED: &str = "Corrupted .tbz2 file";

        expect_magic(&mut file, size - 4, "STOP").context(CORRUPTED)?;
        expect_magic(&mut file, size - 16, "XPAKSTOP").context(CORRUPTED)?;

        let xpak_offset: u64 = u64::from(read_u32(&mut file, size - 8).context(CORRUPTED)?);
        let xpak_start = size - 8 - xpak_offset;

        expect_magic(&mut file, xpak_start, "XPAKPACK").context(CORRUPTED)?;

        let xpak = parse_xpak(&mut file, xpak_start, size)?;

        let category = std::str::from_utf8(
            xpak.get("CATEGORY")
                .with_context(|| "Binary package missing CATEGORY")?,
        )?
        .trim();
        let pf = std::str::from_utf8(
            xpak.get("PF")
                .with_context(|| "Binary package missing PF")?,
        )?
        .trim();
        let category_pf = format!("{category}/{pf}");

        Ok(Self {
            file,
            xpak_start,
            xpak,
            category_pf,
        })
    }

    /// Returns the XPAK key-value map.
    pub fn xpak(&self) -> &HashMap<String, Vec<u8>> {
        &self.xpak
    }

    /// Returns the string combining CATEGORY and PF, e.g. "sys-apps/attr-2.5.1".
    pub fn category_pf(&self) -> &str {
        &self.category_pf
    }

    /// Returns a tarball reader.
    pub fn new_tarball_reader(&mut self) -> Result<std::io::Take<&mut File>> {
        self.file.rewind()?;
        Ok((&mut self.file).take(self.xpak_start))
    }
}

fn read_u32(f: &mut File, offset: u64) -> Result<u32> {
    f.seek(Start(offset))?;
    let mut buffer = [0_u8; std::mem::size_of::<u32>()];
    f.read_exact(&mut buffer)?;
    Ok(bytes::BigEndian::read_u32(&buffer))
}

fn expect_magic(f: &mut File, offset: u64, want: &str) -> Result<()> {
    f.seek(Start(offset))?;
    let mut got: String = "".to_string();
    f.take(want.len() as u64).read_to_string(&mut got)?;
    if got != want {
        bail!("Bad magic: got {got}, want {want}");
    }
    Ok(())
}

fn parse_xpak(file: &mut File, xpak_start: u64, size: u64) -> Result<HashMap<String, Vec<u8>>> {
    let index_len = u64::from(read_u32(file, xpak_start + 8)?);
    let data_len = u64::from(read_u32(file, xpak_start + 12)?);
    let index_start = xpak_start + 16;
    let data_start = index_start + index_len;
    if data_start + data_len != size - 16 {
        bail!("corrupted .tbz2 file: data length inconsistency")
    }

    let mut xpak: HashMap<String, Vec<u8>> = HashMap::new();
    let mut index_pos = index_start;
    while index_pos < data_start {
        let name_len = u64::from(read_u32(file, index_pos)?);
        index_pos += 4;
        let mut name: String = String::new();
        file.take(name_len).read_to_string(&mut name)?;
        if name.len() != name_len.try_into()? {
            bail!("Got '{name}', want a name of length {name_len}")
        }
        index_pos += name_len;
        let data_offset = u64::from(read_u32(file, index_pos)?);
        index_pos += 4;
        let data_len = u64::from(read_u32(file, index_pos)?);
        index_pos += 4;

        file.seek(Start(data_start + data_offset))?;
        let mut data = Vec::new();
        file.take(data_len).read_to_end(&mut data)?;
        if data.len() != data_len.try_into()? {
            bail!(
                "Got a buffer of length {}, want length {}",
                data.len(),
                data_len
            );
        }

        xpak.insert(name, data);
    }
    Ok(xpak)
}

#[cfg(test)]
mod tests {
    use super::*;

    const BINARY_PKG_RUNFILE: &str =
        "cros/bazel/ebuild/private/common/portage/binarypackage/testdata/nano.tbz2";

    // This is the contents of the tarball from the binary package, extracted with
    // qtbz2. We are unable to use the qtbz2 binary in bazel, as this would require
    // first being able to extract the qtbz2 binary from a package.
    const TARBALL_RUNFILE: &str =
        "cros/bazel/ebuild/private/common/portage/binarypackage/testdata/nano.tar.bz2";

    fn binary_package() -> Result<BinaryPackage> {
        let r = runfiles::Runfiles::create()?;
        BinaryPackage::open(&r.rlocation(BINARY_PKG_RUNFILE))
    }

    #[test]
    fn xpak() -> Result<()> {
        let bp = binary_package()?;
        let xpak = bp.xpak();
        assert_eq!(
            xpak.get("CATEGORY")
                .map(|x| std::str::from_utf8(x).unwrap()),
            Some("app-editors\n")
        );
        assert_eq!(
            xpak.get("PF").map(|x| std::str::from_utf8(x).unwrap()),
            Some("nano-6.4\n")
        );
        assert_eq!(
            xpak.get("repository")
                .map(|x| std::str::from_utf8(x).unwrap()),
            Some("portage-stable\n")
        );
        Ok(())
    }

    #[test]
    fn category_pf() -> Result<()> {
        let bp = binary_package()?;
        assert_eq!("app-editors/nano-6.4", bp.category_pf());
        Ok(())
    }

    #[test]
    fn valid_tarball() -> Result<()> {
        let r = runfiles::Runfiles::create()?;
        let mut bp = binary_package()?;

        let mut got: Vec<u8> = Vec::new();
        bp.new_tarball_reader()?.read_to_end(&mut got)?;

        let mut want: Vec<u8> = Vec::new();
        File::open(r.rlocation(TARBALL_RUNFILE))?.read_to_end(&mut want)?;

        assert_eq!(got.len(), want.len());
        // Don't use assert_eq, since the files are massive and it'd print out bytes to stderr.
        assert!(got == want);

        Ok(())
    }
}
